File: app/src/main/java/tipz/browservio/tabbies/CustomTabsActivity.java
package tipz.browservio.tabbies;

import android.content.Intent;
import android.net.Uri;
import android.os.Bundle;
import android.widget.ProgressBar;

import androidx.appcompat.widget.AppCompatImageView;
import androidx.appcompat.widget.AppCompatTextView;
import androidx.swiperefreshlayout.widget.SwipeRefreshLayout;

import tipz.browservio.R;
import tipz.browservio.utils.CommonUtils;
import tipz.browservio.utils.UrlUtils;
import tipz.browservio.webview.VioWebView;
import tipz.browservio.webview.VioWebViewActivity;

public class CustomTabsActivity extends VioWebViewActivity {
    private AppCompatTextView title;
    private AppCompatTextView host;
    private VioWebView webview;

    @Override
    protected void onCreate(Bundle _savedInstanceState) {
        super.onCreate(_savedInstanceState);
        setContentView(R.layout.custom_tab);

        /* Back button */
        AppCompatImageView actionBarBack = findViewById(R.id.actionBarBack);
        actionBarBack.setOnClickListener(v -> finish());

        /* Title and Host */
        title = findViewById(R.id.title);
        host = findViewById(R.id.host);

        /* Share */
        AppCompatImageView share = findViewById(R.id.share);
        share.setOnClickListener(v -> CommonUtils.shareUrl(this, webview.getUrl()));

        /* Open in Browser */
        AppCompatImageView open_browser = findViewById(R.id.open_browser);
        open_browser.setOnClickListener(v -> {
            Intent intent = new Intent(CustomTabsActivity.this, BrowserActivity.class);
            intent.setData(Uri.parse(UrlUtils.cve_2017_13274(webview.getUrl())));
            startActivity(intent);
            finish();
        });

        /* Swipe Refresh Layout */
        SwipeRefreshLayout swipeRefreshLayout = findViewById(R.id.swipe);

        /* Progress Bar */
        ProgressBar MainProg = findViewById(R.id.MainProg);

        /* WebView */
        webview = findViewById(R.id.webview);
        webview.doSettingsCheck();
        webview.setUpProgressBar(MainProg);
        webview.setUpSwipeRefreshLayout(swipeRefreshLayout);
        webview.setUpdateHistory(false);
        Uri dataUri = getIntent().getData();
        if (dataUri != null)
            webview.loadUrl(dataUri.toString());
    }

    @Override
    public void onUrlUpdated(String url) {
        this.host.setText(Uri.parse(UrlUtils.cve_2017_13274(url)).getHost());
    }

    @Override
    public void onTitleUpdated(String title) {
        this.title.setText(title);
    }
}

File: app/src/main/java/tipz/viola/webview/activity/CustomTabsActivity.kt
// Copyright (c) 2022-2025 Tipz Team
// SPDX-License-Identifier: Apache-2.0

package tipz.viola.webview.activity

import android.content.Intent
import android.net.Uri
import android.os.Bundle
import android.view.View
import androidx.appcompat.widget.AppCompatImageView
import androidx.appcompat.widget.AppCompatTextView
import tipz.viola.databinding.ActivityCustomTabBinding
import tipz.viola.ext.finishAndRemoveTaskExt
import tipz.viola.ext.shareUrl
import tipz.viola.webview.VWebViewActivity
import androidx.core.net.toUri

class CustomTabsActivity : VWebViewActivity(true) {
    private lateinit var binding: ActivityCustomTabBinding
    private lateinit var title: AppCompatTextView
    private lateinit var host: AppCompatTextView

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = ActivityCustomTabBinding.inflate(layoutInflater)
        val view = binding.root
        setContentView(view)

        /* Appbar */
        appbar = binding.appbar
        webviewContainer = binding.webviewContainer

        /* Back button */
        val actionBarBack : AppCompatImageView = binding.close
        actionBarBack.setOnClickListener { finishAndRemoveTaskExt() }

        /* Title and Host */
        title = binding.title
        host = binding.host

        /* Share */
        val share = binding.share
        share.setOnClickListener { shareUrl(webview.url) }

        /* Open in Browser */
        val openBrowser = binding.openBrowser
        openBrowser.setOnClickListener {
            val url = webview.url
            val intent = Intent(this, BrowserActivity::class.java)
            intent.data = url.toUri()
            startActivity(intent)
            finishAndRemoveTaskExt()
        }

        /* Progress Bar */
        progressBar = binding.webviewProgressBar

        /* Swipe Refresh Layout */
        swipeRefreshLayout = binding.layoutWebview.swipe

        // Setup Web App Mode
        if (intent.getBooleanExtra(EXTRA_LAUNCH_AS_WEBAPP, false))
            appbar.visibility = View.GONE

        /* WebView */
        webview = binding.layoutWebview.webview
        webview.setUpdateHistory(false)

        // Finally, load webpage
        super.onStart()
        val dataUri = intent.data
        if (dataUri != null) webview.loadUrl(dataUri.toString())
    }

    override fun onUrlUpdated(url: String?) {
        host.text = url!!.toUri().host
    }

    override fun onTitleUpdated(title: String?) {
        super.onTitleUpdated(title)
        this.title.text = title
    }

    companion object {
        const val EXTRA_LAUNCH_AS_WEBAPP = "launchAsWebApp"
    }
}
